src/newton_functs.cpp

```diff
@@ -31,6 +31,48 @@ void update_solution (dolfin::Function* iterate, dolfin::Function* update)
     *(iterate->vector()) += *(update_interpolant.vector());
 }
 
+/**
+ * \fn void update_solution_pnp(dolfin::Function* iterate, dolfin::Function* update, double& relative_residual, const Form& L )
+ *
+ * \brief adds the update Function to the current iterate by testing the residual only for the nonlinear PNP equations
+ *              where  L_pnp.CatCat = cationSolution, L_pnp.AnAn = anionSolution, L_pnp.EsEs = potentialSolution.
+ *
+ * \param iterate             Function to be updated
+ * \param update              update Function
+ * \param relative_residual   residual from the previous Newton iteration
+ * \pram L                    RHS of the equations
+ */
+void update_solution_pnp(dolfin::Function* iterate, dolfin::Function* update, double& relative_residual, const Form& L, const dolfin::DirichletBC bc, dolfin::EigenVector* b )
+{
+    double new_residual;
+    int i=1;
+    dolfin::FunctionSpace V( *(iterate->function_space()) );
+    dolfin::Function update_interpolant(V);
+    update_interpolant.interpolate(*update);
+    dolfin::Function _iterate(iterate);
+    *(_iterate->vector()) += *(update_interpolant.vector());
+    L.CatCat = _iterate[0];
+    L.AnAn   = _iterate[1];
+    L.EsEs   = _iterate[2];
+    assemble(*b, L);
+    bc.apply(*b);
+    new_residual = (*b).norm("l2");
+    while ( (new_residual>relative_residual) and i<10)
```

Comment:
we shouldn't hard code parameters like this 10. It should be named ``maxDamps`` for clarity (or some other name that describes it).